### PR TITLE
feat: IFO Vesting

### PR DESCRIPTION
**Pull request by ChefMomota**

### CODE REVIEW

File: src/views/Ifos/components/IfoVesting/VestingPeriod/Claim.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+const ClaimButton: React.FC<Props> = ({ data, fetchUserVestingData }) => {
+  const { t } = useTranslation()
+  const { toastSuccess } = useToast()
+  const { address, token } = data.ifo
+  const { vestingId } = data.userVestingData[PoolIds.poolUnlimited]
+  const contract = useIfoV3Contract(address)
+  const { fetchWithCatchTxError, loading: isPending } = useCatchTxError()
+
+  const handleClaim = async () => {
```
Comment (chefilip) [inline src/views/Ifos/components/IfoVesting/VestingPeriod/Claim.tsx]: should we use `useCallback`?

Comment (ChefMomota) [reply src/views/Ifos/components/IfoVesting/VestingPeriod/Claim.tsx]: Should we use? Cause i see Pool & farm not using `useCallback `

Comment (chefilip) [reply src/views/Ifos/components/IfoVesting/VestingPeriod/Claim.tsx]: For this case, it would not impact much but it's a good practice. Your call.

Comment (ChefMomota) [reply src/views/Ifos/components/IfoVesting/VestingPeriod/Claim.tsx]: Updated.

**Comment by chefilip** [comment]
LGTM

**Review by chefilip — APPROVED** [review]

File: src/views/Ifos/hooks/vesting/useFetchVestingData.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  userDataLoaded: false,
+}
+
+const useFetchVestingData = (account: string) => {
+  const [userVestingData, setUserVestingData] = useState<VestingDataState>(initialData)
+
+  const fetchUserVestingData = useCallback(async () => {
+    if (account) {
+      const data = await Promise.all(
+        await allVestingIfo.map(async (ifo) => {
```
Comment (0xjojoex) [inline src/views/Ifos/hooks/vesting/useFetchVestingData.ts]: don't need first `await`, and consider use swr for this

File: src/views/Ifos/hooks/v3/useGetPublicIfoData.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+          },
+        ].filter(Boolean),
+      )
+
+      const poolBasicFormatted = formatPool(poolBasic)
+      const poolUnlimitedFormatted = formatPool(poolUnlimited)
+
+      const startBlockNum = startBlock ? startBlock[0].toNumber() : 0
+      const endBlockNum = endBlock ? endBlock[0].toNumber() : 0
+      const taxRateNum = taxRate ? FixedNumber.from(taxRate[0]).divUnsafe(TAX_PRECISION).toUnsafeFloat() : 0
```
Comment (0xjojoex) [inline src/views/Ifos/hooks/v3/useGetPublicIfoData.ts]: i'm partially ok with using bignumberjs for new feature, but why mixing both bignumber.js and ethers?

File: src/views/Ifos/hooks/v3/useGetWalletIfoData.ts [hunk 4]
```diff
@@ -74,20 +81,68 @@ const useGetWalletIfoData = (ifo: Ifo): WalletIfoData => {
       params: [account, [0, 1]],
     }))
 
+    const IfoV3Contract = getIfoV3Contract(address)
```
Comment (chef-jojo) [inline src/views/Ifos/hooks/v3/useGetWalletIfoData.ts:84]: same contract?

Comment (0xjojoex) [inline src/views/Ifos/hooks/v3/useGetWalletIfoData.ts]: `const contract = useIfoV3Contract(address)`
same contract above?

Comment (ChefMomota) [reply src/views/Ifos/hooks/v3/useGetWalletIfoData.ts:84]: Ya, is same. But i can't get read data from `const contract = useIfoV3Contract(address)`

File: src/views/Ifos/hooks/vesting/useFetchVestingData.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import useSWR from 'swr'
+import { useWeb3React } from '@web3-react/core'
+import { Ifo, PoolIds } from 'config/constants/types'
+import { ifosConfig, FAST_INTERVAL } from 'config/constants'
+import { fetchUserWalletIfoData } from './fetchUserWalletIfoData'
+
+// Filter Ifo when isActive = true
+const allVestingIfo: Ifo[] = ifosConfig.filter((ifo) => ifo.version >= 3.2 && ifo.vestingTitle)
```
Comment (ChefMomota) [inline src/views/Ifos/hooks/vesting/useFetchVestingData.ts]: // TODO: revert back before merge.
const allVestingIfo: Ifo[] = ifosConfig.filter((ifo) => ifo.version >= 3.2 && ifo.vestingTitle && !ifo.isActive)

**Pull request merged**